quiche/src/lib.rs

```diff
@@ -7961,13 +8134,143 @@ impl TransportParams {
                 ),
                 initial_max_streams_bidi: Some(self.initial_max_streams_bidi),
                 initial_max_streams_uni: Some(self.initial_max_streams_uni),
-
-                preferred_address: None,
+                preferred_address,
             },
         )
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+struct PreferredAddressParams {
+    addr_v4: Option<SocketAddrV4>,
+    addr_v6: Option<SocketAddrV6>,
+    connection_id: ConnectionId<'static>,
+    stateless_reset_token: u128,
+}
+
+impl PreferredAddressParams {
+    fn new(
+        addr_v4: Option<SocketAddrV4>, addr_v6: Option<SocketAddrV6>,
+        connection_id: ConnectionId<'static>, stateless_reset_token: u128,
+    ) -> Self {
+        PreferredAddressParams {
+            addr_v4,
+            addr_v6,
+            connection_id,
+            stateless_reset_token,
+        }
+    }
+
+    fn decode(b: &mut octets::Octets) -> Result<Self> {
+        let addr_v4 = PreferredAddressParams::decode_v4(b)?;
+        let addr_v6 = PreferredAddressParams::decode_v6(b)?;
+
+        if addr_v4.is_none() && addr_v6.is_none() {
+            return Err(Error::InvalidTransportParam);
+        }
+
+        let cid = b.get_bytes_with_varint_length()?.to_vec().into();
+        let stateless_reset_token = u128::from_be_bytes(
+            b.get_bytes(16)?
+                .to_vec()
+                .try_into()
+                .unwrap(), // Since a 128 bit vector has just successfully been created above, this should never panic.
+        );
+
+        Ok(PreferredAddressParams::new(
+            addr_v4,
+            addr_v6,
+            cid,
+            stateless_reset_token,
+        ))
+    }
+
+    fn decode_v4(b: &mut octets::Octets) -> Result<Option<SocketAddrV4>> {
+        let mut addr_v4 = None;
+        let ip_v4_b = <[u8; 4]>::try_from(b.get_bytes(4)?.slice(4)?)
+            .map_err(|_| Error::InvalidTransportParam)?;
+        let ip_v4 = Ipv4Addr::from(ip_v4_b);
```

Comment:
This might be clearer by just calling b.slice(4)? directly like 

```
let ip_v4 = Ipv4Addr::from(b.slice(4));
b.advance(4);
```

but with some actual error handling and mapping.